semantic-core/src/Analysis/Eval.hs

```diff
@@ -70,147 +78,140 @@ eval Analysis{..} eval = \case
-  , ann (pure "(NilClass)" Core.... block
-    [ ann (Core.edge Import (pure "Class"))
-    , ann (Core.edge Import (pure "(Object)"))
-    ])
-  , ann (let' "NilClass" .= Core.frame)
-  , ann (pure "NilClass" Core.... block
-    [ ann (Core.edge Import (pure "(NilClass)"))
-    , ann (Core.edge Import (pure "Object"))
-    , ann (let' "nil?" .= lam' "_" true)
-    , ann (let' __semantic_truthy .= lam' "_" (Core.bool False))
-    ])
-
-  , ann (let' "(TrueClass)" .= Core.frame)
-  , ann (pure "(TrueClass)" Core.... block
-    [ ann (Core.edge Import (pure "Class"))
-    , ann (Core.edge Import (pure "(Object)"))
-    ])
-  , ann (let' "TrueClass" .= Core.frame)
-  , ann (pure "TrueClass" Core.... block
-    [ ann (Core.edge Import (pure "(TrueClass)"))
-    , ann (Core.edge Import (pure "Object"))
+    , Core.record [ ("thing", pure "dep" Core.... "var") ]
     ])
-
-  , ann (let' "(FalseClass)" .= Core.frame)
-  , ann (pure "(FalseClass)" Core.... block
-    [ ann (Core.edge Import (pure "Class"))
-    , ann (Core.edge Import (pure "(Object)"))
-    ])
-  , ann (let' "FalseClass" .= Core.frame)
-  , ann (pure "FalseClass" Core.... block
-    [ ann (Core.edge Import (pure "(FalseClass)"))
-    , ann (Core.edge Import (pure "Object"))
-    , ann (let' __semantic_truthy .= lam' "_" (Core.bool False))
-    ])
-
-  , ann (let' "nil"   .= pure "NilClass"   $$$ "new")
-  , ann (let' "true"  .= pure "TrueClass"  $$$ "new")
-  , ann (let' "false" .= pure "FalseClass" $$$ "new")
-
-  , ann (let' "require" .= lam' "path" (Core.load (pure "path")))
   ]
-  where -- _nil  = pure "nil"
-        true  = pure "true"
-        false = pure "false"
-        self $$$ method = annWith callStack $ lam' "_x" (pure "_x" Core.... pure method $$ pure "_x") $$ self
 
+ruby :: File (Term Core User)
+ruby = fromBody $ annWith callStack (rec (named' __semantic_global) (do' statements))
+  where statements =
+          [ Just "Class" :<- record
+            [ (__semantic_super, Core.record [])
+            , ("new", lam "self"
+              (    "instance" :<- record [ (__semantic_super, var "self") ]
+              >>>= var "instance" $$$ "initialize"))
+            ]
+
+          , Just "(Object)" :<- record [ (__semantic_super, var "Class") ]
+          , Just "Object" :<- record
+            [ (__semantic_super, var "(Object)")
+            , ("nil?", lam "_" (var __semantic_global ... "false"))
+            , ("initialize", lam "self" (var "self"))
+            , (__semantic_truthy, lam "_" (bool True))
+            ]
+
+          , Just "(NilClass)" :<- record
+            -- FIXME: what should we do about multiple import edges like this
+            [ (__semantic_super, var "Class")
+            , (__semantic_super, var "(Object)")
+            ]
+          , Just "NilClass" :<- record
+            [ (__semantic_super, var "(NilClass)")
+            , (__semantic_super, var "Object")
+            , ("nil?", lam "_" (var __semantic_global ... "true"))
```

Comment:
We look `true` up in `__semantic_global` (which is bound recursively) because it hasn’t been defined yet.